### PR TITLE
goyacc: don't import "unsafe" if not used

### LINKED ISSUE #13006: Bug Report: generated parsers import "unsafe" even if they are not using "fast append"
### Overview of the Issue

I'm using the fork of `goyacc` in `go/vt/sqlparser/goyacc`. It has an optimization that is enabled via a command-line flag for faster appending of slices (reduces overhead of storing slices in interface values, which ends up allocating a new small space on the heap for the slice header).

I am not actually using this flag, so my generated parser makes no references to the generated `*Iaddr` function. But the generated parser still _contains_ this function definition, unused, and also must import `"unsafe"`. For maximum portability of generated parsers (like if targeting runtimes that do not support `"unsafe"`), I think this import and the `*Iaddr` function should only be included in the output when the fast-append flag is set.

### Reproduction Steps

Create a `example.y` grammar and compile it with the `goyacc` fork in this repo. Do not use the `-fast-append` command-line flag. The generated parser `example.y.go` file will import `"unsafe"`, but it is only referenced from an unused function named `exampleIaddr`.

### Binary Version

```sh
N/A
```

### Operating System and Environment details

```sh
I'm using a MBP (M1 Max) with OS X 13.3.1 (Ventura).
```

### Log Fragments

_No response_

### PR DESCRIPTION
**Pull request by jhump**
## Description

Avoids importing "unsafe" (and emitting the `*Iaddr` helper function that uses it) if the `--fast-append` flag is not used. When this flag is not used, the `*Iaddr` helper function is never used. To maximize code portability (since there are some niche runtimes that do not support unsafe), we want to avoid the "unsafe" import if we can.

## Related Issue(s)

Fixes #13006 

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [ ] Documentation was added or is not required

I did some manual testing with a yacc grammar I have. The change excluded the import and the unused helper function as expected. When I added the `-f` flag, they were present (along with a few calls to the `*Iaddr` function).

This still leaves some potential room for improvement: instead of gating this only on the `--fast-append` flag, we could also see if the token type actually includes any slice fields. If not, we could omit this. Counter-argument: it might be weird if the "unsafe" import magically appears just from adding a new field to the union type in the grammar. 🤷 

## Deployment Notes

N/A

### CODE REVIEW

**Review by GuptaManan100 — COMMENTED** [review]
Everything else looks okay to me, however you need to fix the `sql.go` file by regenerating it. These changes cause a slight change in the file (a comment is reordered).

**Comment by jhump** [comment]
>Everything else looks okay to me, however you need to fix the sql.go file by regenerating it. These changes cause a slight change in the file (a comment is reordered).

@GuptaManan100, done. I ran `go install golang.org/x/tools/cmd/goimports && make codegen`.

**Review by GuptaManan100 — APPROVED** [review]
LGTM now.

**Review by vmg — APPROVED** [review]
Seems reasonable. @jhump: the static linter doesn't like the `fmt.Fprintf` call in line 3054.

File: go/vt/sqlparser/goyacc/goyacc.go [hunk 1]
```diff
@@ -3047,6 +3049,11 @@ func others() {
 		ch = getrune(finput)
 	}
 
+	if allowFastAppend {
+		fastAppendHelper := strings.Replace(fastAppendHelperText, "$$", prefix, -1)
+		fmt.Fprintf(ftable, fastAppendHelper)
```
Comment (jhump) [inline go/vt/sqlparser/goyacc/goyacc.go:3054]:
```suggestion
		fmt.Fprint(ftable, fastAppendHelper)
```

File: go/vt/sqlparser/goyacc/goyacc.go [hunk 2]
```diff
@@ -3047,6 +3049,11 @@ func others() {
...
+	}
+
 	// copy yaccpar
 	if !lflag {
 		fmt.Fprintf(ftable, "\n//line yaccpar:1\n")
```
Comment (jhump) [inline go/vt/sqlparser/goyacc/goyacc.go:3059]:
```suggestion

	// copy yaccpar
	if !lflag {
		fmt.Fprint(ftable, "\n//line yaccpar:1\n")
```

**Comment by jhump** [comment]
> Seems reasonable. @jhump: the static linter doesn't like the `fmt.Fprintf` call in line 3054.

@vmg, fixed.

**Pull request merged**